pennylane/templates/subroutines/select.py

```diff
@@ -0,0 +1,162 @@
+from pennylane import math
+
+
+class Select(Operation):
+    r"""Applies specific input operations depending on the state of
+    the designated control qubits.
+
+    .. math:: Select|i\rangle \otimes |\psi\rangle = |i\rangle \otimes U_i |\psi\rangle
+
+    Args:
+        ops (list[Operator]): operations to apply
+        control_wires (Sequence[int]): the wires controlling which operation is applied
+        id (str or None): String representing the operation (optional)
+
+    .. note::
+        The position of the operation in the list determines which qubit state implements that operation.
+        For example, when the qubit register is in the state :math:`|00\rangle`, we will apply ``ops[0]``.
+        When the qubit register is in the state :math:`|10\rangle`, we will apply ``ops[2]``. To obtain the
+        binary bitstring representing the state for list position ``index`` we can use the following relationship:
+        ``index = int(state_string, 2)``. For example, ``2 = int('10', 2)``.
+
+    **Example**
+
+    >>> dev = qml.device('default.qubit',wires=4)
+    >>> ops = [qml.PauliX(wires=2),qml.PauliX(wires=3),qml.PauliY(wires=2),qml.SWAP([2,3])]
+    >>> @qml.qnode(dev)
+    >>> def circuit():
+    >>>     qml.Select(ops,control_wires=[0,1])
+    >>>     return qml.state()
+    ...
+    >>> print(qml.draw(circuit,expansion_strategy='device')())
+    0: ─╭○─╭○─╭●─╭●────┤  State
+    1: ─├○─├●─├○─├●────┤  State
+    2: ─╰X─│──╰Y─├SWAP─┤  State
+    3: ────╰X────╰SWAP─┤  State
+
+    """
+
+    num_wires = qml.operation.AnyWires
+
+    def _flatten(self):
+        return (self.ops), (self.control_wires)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "Select":
+        return cls(data, metadata)
+
+    def __init__(self, ops, control_wires, id=None):
+        control_wires = qml.wires.Wires(control_wires)
+        self.hyperparameters["ops"] = ops
+        self.hyperparameters["control_wires"] = control_wires
+
+        if 2 ** len(control_wires) < len(ops):
+            raise ValueError(
+                f"Not enough control wires ({len(control_wires)}) for the desired number of "
+                + f"operations ({len(ops)}). At least {int(math.ceil(math.log2(len(ops))))} control "
+                + "wires required."
+            )
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control_wires
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+        for op in ops:
+            qml.QueuingManager.remove(op)
+
+        target_wires = qml.wires.Wires.all_wires([op.wires for op in ops])
+        self.hyperparameters["target_wires"] = target_wires
+
+        all_wires = target_wires + control_wires
+        super().__init__(ops, wires=all_wires, id=id)
+
+    def decomposition(self):
+        r"""Representation of the operator as a product of other operators.
```

Comment:
Can we add an example for the decomposition in this docstring or in `compute_decomposition`?

Reply:
Added in the next push